src/wrapper.rs

```diff
@@ -93,102 +93,148 @@ impl DxcOperationResult {
-                (*me).pinned.push(source);
-            }
+                unsafe { *include_source = Some(DxcBlob::from(blob).inner) };
+                self.pinned.borrow_mut().push(source);
 
-            0
-        } else {
-            -2_147_024_894 // ERROR_FILE_NOT_FOUND / 0x80070002
+                // NOERROR
+                0
+            } else {
+                -2_147_024_894 // ERROR_FILE_NOT_FOUND / 0x80070002
+            }
+            .into()
         }
-        .into()
     }
 }
 
-#[derive(Debug)]
+/// Represents a reference to a COM object that should only live as long as itself
+///
+/// In other words, on [`drop()`] we should be able to decrement the refcount to zero.
+/// This object functions a lot like [`ClassAllocation`]: see its similar [`drop()`]
+/// implementation for details.
+///
+/// Note that COM objects live on the heap by design, because of this refcount system.
+struct LocalClassAllocation<T: Class>(core::pin::Pin<Box<T>>);
+
+impl<T: Class> LocalClassAllocation<T> {
+    fn new(allocation: ClassAllocation<T>) -> Self {
+        // TODO: There is no way to take the internal, owned box out of com-rs's
+        // allocation wrapper.
+        // https://github.com/microsoft/com-rs/issues/236 covers this issue as a whole,
+        // including lifetime support and this `LocalClassAllocation` upstream.
+        let inner: core::mem::ManuallyDrop<core::pin::Pin<Box<T>>> =
+            unsafe { std::mem::transmute(allocation) };
+
+        Self(core::mem::ManuallyDrop::into_inner(inner))
+    }
+
+    // TODO: Return a borrow of this interface?
+    // query_interface() is not behind one of the traits
+    // fn query_interface<T>(&self) -> Option<T> {
+    //     self.0.query_interface::<T>().unwrap()
+    // }
+}
+
+impl<T: Class> Deref for LocalClassAllocation<T> {
+    type Target = core::pin::Pin<Box<T>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T: Class> Drop for LocalClassAllocation<T> {
+    fn drop(&mut self) {
+        // Check if we are the only remaining reference to this object
+        assert_eq!(
+            unsafe { self.0.dec_ref_count() },
+            0,
+            "COM object is still referenced"
+        );
+        // Now that we're the last one to give up our refcount, it is safe
+        // for the internal object to get dropped.
+    }
+}
+
+impl DxcIncludeHandlerWrapper {
+    /// SAFETY: Make sure the returned object does _not_ outlive the lifetime
+    /// of either `library` nor `include_handler`
+    unsafe fn create_include_handler(
+        library: &'_ DxcLibrary,
+        include_handler: &'_ mut dyn DxcIncludeHandler,
+    ) -> LocalClassAllocation<DxcIncludeHandlerWrapper> {
+        LocalClassAllocation::new(Self::allocate(
+            std::mem::transmute(library),
+            RefCell::new(std::mem::transmute(include_handler)),
+            RefCell::new(vec![]),
+        ))
+    }
+}
```

Comment:
"Localized" allocations of COM classes - including lifetime parameter(s) - is discussed upstream at https://github.com/microsoft/com-rs/issues/236.